docs/codeql/codeql-language-guides/analyzing-data-flow-in-ruby.rst

```diff
@@ -0,0 +1,387 @@
+.. _analyzing-data-flow-in-ruby:
+
+Analyzing data flow in Ruby
+=============================
+
+You can use CodeQL to track the flow of data through a Ruby program to places where the data is used.
+
+About this article
+------------------
+
+This article describes how data flow analysis is implemented in the CodeQL libraries for Ruby and includes examples to help you write your own data flow queries.
+The following sections describe how to use the libraries for local data flow, global data flow, and taint tracking.
+For a more general introduction to modeling data flow, see ":ref:`About data flow analysis <about-data-flow-analysis>`."
+
+Local data flow
+---------------
+
+Local data flow is data flow within a single method or callable. Local data flow is easier, faster, and more precise than global data flow, and is sufficient for many queries.
+
+Using local data flow
+~~~~~~~~~~~~~~~~~~~~~
+
+The local data flow library is in the module ``DataFlow`` and it defines the class ``Node``, representing any element through which data can flow.
+``Node``\ s are divided into expression nodes (``ExprNode``) and parameter nodes (``ParameterNode``).
+You can map between a data flow ``ParameterNode`` and its corresponding ``Parameter`` AST node using the ``asParameter`` member predicate.
+Meanwhile, the ``asExpr`` member predicate maps between a data flow ``ExprNode`` and its corresponding ``ExprCfgNode`` in the control-flow library.
+
+.. code-block:: ql
+
+     class Node {
+       /** Gets the expression corresponding to this node, if any. */
+       CfgNodes::ExprCfgNode asExpr() { ... }
+
+       /** Gets the parameter corresponding to this node, if any. */
+       Parameter asParameter() { ... }
+
+      ...
+     }
+
+You can also use the predicates ``exprNode`` and ``parameterNode``:
+
+.. code-block:: ql
+
+     /**
+      * Gets a node corresponding to expression `e`.
+      */
+     ExprNode exprNode(CfgNodes::ExprCfgNode e) { ... }
+
+     /**
+      * Gets the node corresponding to the value of parameter `p` at function entry.
+      */
+     ParameterNode parameterNode(Parameter p) { ... }
+
+Note that since ``asExpr`` and ``exprNode`` map between data-flow and control-flow nodes, you then need to call the ``getExpr`` member predicate on the control-flow node to map to the corresponding AST node,
+e.g. by writing ``node.asExpr().getExpr()``.
+Due to the control-flow graph being split, there can be multiple data-flow and control-flow nodes associated with a single expression AST node.
```

Comment:
> Due to the control-flow graph being split

I'm not sure about this, will suggest an alternative wording, but I may well have misunderstood.

```suggestion
A control-flow graph considers every way control can flow through code, consequently, there can be multiple data-flow and control-flow nodes associated with a single expression node in the AST.
```

Reply:
I think you understood perfectly, thanks.